subsys/net/l2/ieee802154/ieee802154_fragment.c

```diff
@@ -334,34 +345,23 @@ static inline struct frag_cache *get_reass_cache(u16_t size, u16_t tag)
 	return NULL;
 }
 
-/* Helper function to write fragment data to Rx packet based on offset. */
```

Comment:
Seems some confusion here. While receiving 15.4 fragments can come in any order. We should be able to assemble them based on their offset. Its not always appending the buffer. Original code handles that or I am missing something in these changes.

Reply:
Ok, I see. That explains why the test passes: it's fully sequential. I have an idea how this could be fixed, as long as the fragments arrive it would not remove the fragment header, just append them to the net_pkt. For each fragment received it would recompute the offset vs size on the whole list, and if all seems to be fine it would rearrange the list according to offset, and get rid of the fragment hdr on each then.

It would be far less costly than the legacy net_pkt_write (which is doomed to be removed)

Reply:
> as long as the fragments arrive it would not remove the fragment header, just append them to the net_pkt.

Problem here, one fragment header size is 4 bytes (which should be beginning of the packet) and rest of fragments header size is 5 bytes. It's not always same fragment header size to remove later.

Reply:
> Ok, I see. That explains why the test passes: it's fully sequential.

Yep, we don't have random fragment tests.

Reply:
That's ok, the info on the type of header is located on the first byte of the buffer, so header removal would take that into account to know the size of hdr to remove.

Reply:
Yes